File: backend/models/Room.js
const mongoose = require('mongoose');

const roomSchema = new mongoose.Schema({
  name: {
    type: String,
    required: [true, 'Room name is required'],
    trim: true,
    maxlength: [100, 'Room name cannot exceed 100 characters']
  },
  roomNumber: {
    type: String,
    required: [true, 'Room number is required'],
    unique: true,
    trim: true
  },
  category: {
    type: String,
    required: [true, 'Room category is required'],
    enum: ['standard', 'executive', 'suite', 'family', 'presidential'],
    lowercase: true
  },
  description: {
    type: String,
    required: [true, 'Room description is required'],
    maxlength: [1000, 'Description cannot exceed 1000 characters']
  },
  price: {
    type: Number,
    required: [true, 'Room price is required'],
    min: [0, 'Price cannot be negative']
  },
  originalPrice: {
    type: Number,
    min: [0, 'Original price cannot be negative']
  },
  size: {
    type: String,
    required: [true, 'Room size is required']
  },
  maxGuests: {
    type: Number,
    required: [true, 'Maximum guests is required'],
    min: [1, 'Maximum guests must be at least 1'],
    max: [10, 'Maximum guests cannot exceed 10']
  },
  bedConfiguration: {
    type: String,
    required: [true, 'Bed configuration is required']
  },
  images: [{
    url: {
      type: String,
      required: true
    },
    alt: {
      type: String,
      default: ''
    },
    isPrimary: {
      type: Boolean,
      default: false
    }
  }],
  amenities: [{
    name: {
      type: String,
      required: true
    },
    icon: String,
    category: {
      type: String,
      enum: ['basic', 'entertainment', 'bathroom', 'comfort', 'technology']
    }
  }],
  features: [{
    type: String,
    trim: true
  }],
  detailedAmenities: {
    popularWithGuests: [String],
    roomFeatures: [String],
    mediaEntertainment: [String],
    bathroom: [String],
    otherFacilities: [String]
  },
  floor: {
    type: Number,
    required: [true, 'Floor number is required'],
    min: [1, 'Floor must be at least 1']
  },
  view: {
    type: String,
    enum: ['city', 'ocean', 'garden', 'pool', 'courtyard'],
    default: 'city'
  },
  status: {
    type: String,
    enum: ['available', 'occupied', 'maintenance', 'cleaning', 'out-of-order'],
    default: 'available'
  },
  isActive: {
    type: Boolean,
    default: true
  },
  rating: {
    average: {
      type: Number,
      default: 0,
      min: 0,
      max: 5
    },
    count: {
      type: Number,
      default: 0
    }
  },
  reviews: [{
    user: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: true
    },
    rating: {
      type: Number,
      required: true,
      min: 1,
      max: 5
    },
    comment: {
      type: String,
      maxlength: [500, 'Review comment cannot exceed 500 characters']
    },
    date: {
      type: Date,
      default: Date.now
    }
  }],
  popular: {
    type: Boolean,
    default: false
  },
  bookingCount: {
    type: Number,
    default: 0
  },
  lastCleaned: {
    type: Date
  },
  lastMaintenance: {
    type: Date
  }
}, {
  timestamps: true
});

// Indexes for better query performance
roomSchema.index({ category: 1, status: 1 });
roomSchema.index({ price: 1 });
roomSchema.index({ 'rating.average': -1 });
roomSchema.index({ popular: -1, 'rating.average': -1 });

// Calculate average rating when reviews are updated
roomSchema.methods.calculateAverageRating = function() {
  if (this.reviews.length === 0) {
    this.rating.average = 0;
    this.rating.count = 0;
  } else {
    const sum = this.reviews.reduce((acc, review) => acc + review.rating, 0);
    this.rating.average = Math.round((sum / this.reviews.length) * 10) / 10;
    this.rating.count = this.reviews.length;
  }
};

// Pre-save middleware to calculate rating
roomSchema.pre('save', function(next) {
  if (this.isModified('reviews')) {
    this.calculateAverageRating();
  }
  next();
});

// Virtual for primary image
roomSchema.virtual('primaryImage').get(function() {
  const primary = this.images.find(img => img.isPrimary);
  return primary ? primary.url : (this.images.length > 0 ? this.images[0].url : null);
});

// Virtual for discount percentage
roomSchema.virtual('discountPercentage').get(function() {
  if (this.originalPrice && this.originalPrice > this.price) {
    return Math.round(((this.originalPrice - this.price) / this.originalPrice) * 100);
  }
  return 0;
});

module.exports = mongoose.model('Room', roomSchema);

File: JSRoomsAPI/JSRoomsAPI/Models/Room.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace JSroomsAPI.Models
{
    [Table("Rooms")]
    public class Room
    {
        [Key]
        public int RoomId { get; set; }

        [Required]
        [StringLength(10)]
        public string RoomNumber { get; set; } = string.Empty;

        [Required]
        [StringLength(200)]
        public string RoomName { get; set; } = string.Empty;

        [Required]
        public int CategoryId { get; set; }

        [Required]
        [StringLength(50)]
        public string CategoryName { get; set; } = string.Empty;

        [StringLength(500)]
        public string Description { get; set; } = string.Empty;

        [Required]
        [Column(TypeName = "decimal(8,2)")]
        public decimal Price { get; set; }

        [Column(TypeName = "decimal(8,2)")]
        public decimal? OriginalPrice { get; set; }

        [StringLength(20)]
        public string Size { get; set; } = string.Empty;

        [Required]
        public int MaxGuests { get; set; } = 1;

        [StringLength(50)]
        public string BedConfiguration { get; set; } = string.Empty;

        public int Floor { get; set; } = 1;

        [StringLength(50)]
        public string View { get; set; } = "City View";

        [StringLength(20)]
        public string Status { get; set; } = "available";

        public bool IsAC { get; set; } = true;

        [StringLength(10)]
        public string AcType { get; set; } = "ac"; // "ac" or "non-ac"

        public bool IsPopular { get; set; } = false;

        [Column(TypeName = "decimal(3,2)")]
        public decimal Rating { get; set; } = 0;

        public int RatingCount { get; set; } = 0;

        [StringLength(1000)]
        public string ImageUrl { get; set; } = string.Empty;

        [StringLength(100)]
        public string ImageAlt { get; set; } = string.Empty;

        [StringLength(500)]
        public string Amenities { get; set; } = string.Empty; // JSON string of amenities

        [Column(TypeName = "date")]
        public DateTime? LastCleaned { get; set; }

        [Column(TypeName = "date")]
        public DateTime? LastMaintenance { get; set; }

        public bool IsActive { get; set; } = true;

        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
        public DateTime? UpdatedAt { get; set; }

        // Navigation properties
        [ForeignKey("CategoryId")]
        public virtual Category? Category { get; set; }

        public virtual ICollection<Booking> Bookings { get; set; } = new List<Booking>();
    }
}